### PR TITLE
Cache sig chain links to speed up loaduser (and thus identify2)

### PR DESCRIPTION
**Pull request by patrickxb**
r? @maxtaco

### CODE REVIEW

File: go/libkb/loaduser.go [hunk 1]
```diff
@@ -113,7 +114,8 @@ func LoadMe(arg LoadUserArg) (*User, error) {
 }
 
 func LoadUser(arg LoadUserArg) (ret *User, err error) {
-	G.Log.Debug("LoadUser: %+v", arg)
+	defer TimeLog(fmt.Sprintf("LoadUser: %+v", arg), time.Now(), arg.G().Log.Debug)
+	arg.G().Log.Debug("LoadUser: %+v", arg)
```
Comment (maxtaco) [inline go/libkb/loaduser.go]: aren't there some redunancies with these debugs and the ones below?

Comment (patrickxb) [inline go/libkb/loaduser.go]: Yes.  All I did on line 118 is use contextified G.

File: go/libkb/sig_chain.go [hunk 2]
```diff
@@ -106,6 +106,16 @@ func last(links []*ChainLink) (ret *ChainLink) {
 	return links[len(links)-1]
 }
 
+func (sc *SigChain) ChainLinksCopy() []ChainLink {
+	links := make([]ChainLink, len(sc.chainLinks))
+	for i := 0; i < len(sc.chainLinks); i++ {
+		links[i] = *sc.chainLinks[i]
```
Comment (maxtaco) [inline go/libkb/sig_chain.go]: Does this need to be a deep copy?  `ChainLink` itself has a bunch of pointers dangling off of it.

Comment (maxtaco) [inline go/libkb/sig_chain.go]: Can we change `SigChain` to have a list of `ChainLinks` rather than a list of `*ChainLink`s?

**Comment by patrickxb** [comment]
@maxtaco PTAL

**Comment by maxtaco** [comment]
will do after bedtime!

**Comment by maxtaco** [comment]
:+1: This is great, so simple.  Low-hanging fruit: plucked?

File: go/libkb/link_cache.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+// this source code is governed by the included BSD license.
+
+package libkb
+
+type linkIDFixed [LinkIDLen]byte
+
+// LinkCache is a cache of ChainLinks.
+// It is safe to use concurrently.
+type LinkCache struct {
+	cache map[linkIDFixed]ChainLink
```
Comment (maxtaco) [inline go/libkb/link_cache.go]: Hmm, maybe we don't want this thing growing without bound.  Can we periodically clean out links we haven't used in a while past a certain buffer size?

File: go/libkb/link_cache.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		dels:  make(chan delLinkReq),
+		done:  make(chan struct{}),
+	}
+
+	go c.handle()
+
+	return c
+}
+
+// Get retrieves a list of ChainLinks from the cache. If nothing
```
Comment (maxtaco) [inline go/libkb/link_cache.go]: comment here is out-of-date

File: go/libkb/link_cache.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+		result: make(chan getLinkRes),
+	}
+	copy(req.linkID[:], id)
+	c.gets <- req
+
+	res := <-req.result
+	return res.link, res.ok
+}
+
+// Put inserts a list of ChainLinks into the cache.
```
Comment (maxtaco) [inline go/libkb/link_cache.go]: stale comment

File: go/libkb/link_cache.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+func (c *LinkCache) Put(id LinkID, link ChainLink) {
+	req := putLinkReq{
+		link: link,
+	}
+	copy(req.linkID[:], id)
+	c.puts <- req
+}
+
+// Remove deletes a list of ChainLinks from the cache.
```
Comment (maxtaco) [inline go/libkb/link_cache.go]: stale comment

**Comment by maxtaco** [comment]
(On second read there were a few small things worth addressing... But no need for a second review, looks good!)

**Comment by maxtaco** [comment]
Sorry for the lag.  I'm not happy with the cache cleaning strategy.  I know keys ordering is random in Go, but random with which parameters?  I worry the code will always clean elements that will always be directly reinserted.  We might just be churning like crazy.

**Comment by maxtaco** [comment]
Can we use Hashicorp's LRU for this inside the main thread?

**Comment by maxtaco** [comment]
I realize we'd be locking unnecessarily, but I think it's fine for now...

**Comment by patrickxb** [comment]
We've used stathat's ramcache elsewhere for LRU.

**Comment by maxtaco** [comment]
that's good too, either/or

**Comment by patrickxb** [comment]
I modified this to do LRU eviction.

If you prefer, I can replace it with an external package.

**Comment by maxtaco** [comment]
I can take a look tonight.

**Comment by maxtaco** [comment]
Took a look on the subway. Fine be me if the linked list delete on item
deletion is O(1).

**Comment by patrickxb** [comment]
It is:

https://golang.org/src/container/list/list.go?s=2989:3034#L111

**Pull request merged**